vendor/k8s.io/klog/v2/klog.go

```diff
@@ -1223,24 +1224,38 @@ func StartFlushDaemon(interval time.Duration) {
 // lockAndFlushAll is like flushAll but locks l.mu first.
 func (l *loggingT) lockAndFlushAll() {
 	l.mu.Lock()
-	l.flushAll()
+	files := l.flushAll()
 	l.mu.Unlock()
+	// Some environments are slow when syncing and holding the lock might cause contention.
+	l.syncAll(files)
```

Comment:
do we know flushSyncWriter implementations can have Sync() called concurrent with other methods (writes or flushes)?

Reply:
(I can believe the File implementation is ok with Sync being called whenever, but we've defined it as an interface so I don't know what other implementations we have in the mix)

Reply:
That goes back 11 years... my two cents: there has been no guarantee whatsoever whether klog does serialization for that interface, so I expect implementations to be prepared for concurrent calls. I expect them to mimic `os.File`.

I know, hidden implementation details can leak into the API, but how much do we want to be constraint by that?

Reply:
@liggitt i am not aware of any other implementation of this interface other than the one we have in klog. We did inherit this code from glog, so there's an old variant there.

( Do we want to drop the interface @pohly @liggitt ? )

Reply:
> That goes back 11 years... my two cents: there has been no guarantee whatsoever whether klog does serialization for that interface, so I expect implementations to be prepared for concurrent calls. I expect them to mimic os.File.
> 
> I know, hidden implementation details can leak into the API, but how much do we want to be constraint by that?

I'm mostly looking for us to sweep all the implementations linked in for kube binaries, make sure they're safe to call Sync() on concurrent with other methods, and update the docs on the interface to indicate Sync() can be called concurrently.

Reply:
I checked again how that interface is really used. I thought it was public, but it isn't.

There are two implementations of it:
- `redirectBuffer` is used for an `io.Writer` coming from a user of klog, with an empty Sync/Flush method.
- `syncBuffer` is used internally. That is where we really have a problem: the retrieval of `sb.file` is not protected and races with rotating that file.

None of that matters for Kubernetes. Our binaries don't use this code anymore.

But let's fix it first nonetheless.

/hold

Reply:
/hold cancel

v2.130.1 removes these interface because they made it hard to reason about the locking and they weren't really needed.

Reply:
thanks!